[cockroach] Add CockroachDB clustering support to the habitat plan

irvingpop opened the pull request:
This PR enables clustering and also a number of important operational improvements around health checking and logging

Ideas for making this more complete:

- [x]  Enable clustering in "secure" mode

thesentinels commented:
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

predominant commented:
@irvingpop I'm having some issues getting the cluster running inside docker. Here's the log:

```
✗ docker-compose up
Creating network "cockroach_default" with the default driver
Creating volume "cockroach_roach1-data" with default driver
Creating volume "cockroach_roach2-data" with default driver
Creating volume "cockroach_roach3-data" with default driver
Creating cockroach_roach2_1 ... done
Creating cockroach_roach1_1 ... done
Creating cockroach_roach3_1 ... done
Attaching to cockroach_roach2_1, cockroach_roach1_1, cockroach_roach3_1
roach2_1  | hab-sup(MR): core/hab-sup (core/hab-sup/0.65.0/20181004221503)
roach1_1  | hab-sup(MR): core/hab-sup (core/hab-sup/0.65.0/20181004221503)
roach3_1  | hab-sup(MR): core/hab-sup (core/hab-sup/0.65.0/20181004221503)
roach3_1  | hab-sup(MR): Supervisor Member-ID 69c4c158c29f442e9ca4459cad50c804
roach2_1  | hab-sup(MR): Supervisor Member-ID d0423b0b0500461597363ab14dfc08c3
roach2_1  | hab-sup(AG): The rakops/cockroach/2.0.6/20181018065735 service was successfully loaded
roach3_1  | hab-sup(AG): The rakops/cockroach/2.0.6/20181018065735 service was successfully loaded
roach1_1  | hab-sup(MR): Supervisor Member-ID ad17536b5a004471a0f1c51e5f560dcd
roach2_1  | hab-sup(MR): Starting rakops/cockroach/2.0.6/20181018065735 (rakops/cockroach/2.0.6/20181018065735)
roach1_1  | hab-sup(AG): The rakops/cockroach/2.0.6/20181018065735 service was successfully loaded
roach3_1  | hab-sup(MR): Starting rakops/cockroach/2.0.6/20181018065735 (rakops/cockroach/2.0.6/20181018065735)
roach2_1  | cockroach.cluster(UCW): Watching user.toml
roach2_1  | hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
roach1_1  | hab-sup(MR): Starting rakops/cockroach/2.0.6/20181018065735 (rakops/cockroach/2.0.6/20181018065735)
roach1_1  | cockroach.cluster(UCW): Watching user.toml
roach1_1  | hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
roach3_1  | cockroach.cluster(UCW): Watching user.toml
roach3_1  | hab-sup(MR): Starting gossip-listener on 0.0.0.0:9638
roach1_1  | hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
roach2_1  | hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
roach2_1  | hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
roach3_1  | hab-sup(MR): Starting ctl-gateway on 127.0.0.1:9632
roach3_1  | hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
roach2_1  | cockroach.cluster(HK): run, compiled to /hab/svc/cockroach/hooks/run
roach2_1  | cockroach.cluster(HK): Hooks compiled
roach2_1  | cockroach.cluster(SR): Hooks recompiled
roach2_1  | cockroach.cluster(SR): Waiting to execute hooks; election in progress, and we have no quorum.
roach3_1  | cockroach.cluster(HK): run, compiled to /hab/svc/cockroach/hooks/run
roach3_1  | cockroach.cluster(HK): Hooks compiled
roach3_1  | cockroach.cluster(SR): Hooks recompiled
roach3_1  | cockroach.cluster(SR): Waiting to execute hooks; election in progress, and we have no quorum.
roach1_1  | hab-sup(MR): Starting http-gateway on 0.0.0.0:9631
roach1_1  | cockroach.cluster(HK): run, compiled to /hab/svc/cockroach/hooks/run
roach1_1  | cockroach.cluster(HK): Hooks compiled
roach1_1  | cockroach.cluster(SR): Hooks recompiled
roach1_1  | cockroach.cluster(SR): Waiting to execute hooks; election in progress, and we have no quorum.
roach2_1  | cockroach.cluster(HK): Hooks compiled
roach1_1  | cockroach.cluster(HK): Hooks compiled
roach3_1  | cockroach.cluster(HK): Hooks compiled
roach2_1  | cockroach.cluster(HK): Hooks compiled
roach3_1  | cockroach.cluster(HK): Hooks compiled
roach3_1  | cockroach.cluster(HK): Hooks compiled
roach2_1  | cockroach.cluster(HK): Hooks compiled
roach2_1  | cockroach.cluster(SR): Executing hooks; d0423b0b0500461597363ab14dfc08c3 is the leader
roach2_1  | cockroach.cluster(SR): Initializing
roach2_1  | cockroach.cluster(SV): Starting service as user=hab, group=hab
roach1_1  | cockroach.cluster(HK): Hooks compiled
roach1_1  | cockroach.cluster(SR): Executing hooks; d0423b0b0500461597363ab14dfc08c3 is the leader
roach1_1  | cockroach.cluster(SR): Initializing
roach1_1  | cockroach.cluster(SV): Starting service as user=hab, group=hab
roach3_1  | cockroach.cluster(HK): Hooks compiled
roach3_1  | cockroach.cluster(SR): Executing hooks; d0423b0b0500461597363ab14dfc08c3 is the leader
roach3_1  | cockroach.cluster(SR): Initializing
roach3_1  | cockroach.cluster(SV): Starting service as user=hab, group=hab
roach2_1  | cockroach.cluster(O): *
roach2_1  | cockroach.cluster(O): * ERROR: failed to start server: problem using security settings, did you mean to use --insecure?: problem loading certs directory /root/.cockroach-certs: open /root/.cockroach-certs: permission denied
roach2_1  | cockroach.cluster(O): *
roach2_1  | cockroach.cluster(O): Failed running "start"
roach1_1  | cockroach.cluster(O): *
roach1_1  | cockroach.cluster(O): * ERROR: failed to start server: problem using security settings, did you mean to use --insecure?: problem loading certs directory /root/.cockroach-certs: open /root/.cockroach-certs: permission denied
roach1_1  | cockroach.cluster(O): *
roach1_1  | cockroach.cluster(O): Failed running "start"
roach1_1  | hab-launch(SV): Child for service 'cockroach.cluster' with PID 51 exited with code exit code: 1
roach2_1  | hab-launch(SV): Child for service 'cockroach.cluster' with PID 50 exited with code exit code: 1

^C
Killing cockroach_roach1_1  ... done
Killing cockroach_roach3_1  ... done
Killing cockroach_roach2_1  ... done
```

irvingpop commented:
hm, that's really odd @predominant -  --certs-dir is the first option that should get passed in the run hook.   Here's what mine looks like when rendered out (using insecure = false):

```
#!/bin/sh

exec 2>&1
exec cockroach start \
  --certs-dir="/hab/svc/cockroach/var/certs" \
  --join=172.18.0.2:26257 \
  --host=172.18.0.3 \
  --port=26257 \
  --http-host=172.18.0.3 \
  --http-port=8080 \
  --store=/hab/svc/cockroach/data \
  --locality=region=default,datacenter=default \
  --cache=.25 \
  --max-sql-memory=.25 \
  --logtostderr=WARNING \
  --log-dir=/hab/svc/cockroach/var/log \
  --temp-dir=/hab/svc/cockroach/var/tmp
```

Can you check to see what yours looks like?

smacfarlane commented:
@irvingpop sorry this has taken so long to get to.  I had to update the version to 2.1.6 in order to build with gcc8.  When running the docker-compose example though, I was getting the following error:
```
roach1_1  | cockroach.cluster(O): * ERROR: could not record temporary directory path to record file: /hab/svc/cockroach/data/temp-dirs-record.txt: open /hab/svc/cockroach/data/temp-dirs-record.txt: permission denied
```

It looks like the docker-compose volumes used are mounted as `root:root`, while the service runs as `hab`.  I did a quick scan of the docker-compose docs to see if there was a way to mount as a different user and wasn't able to find one.  

Removing the volume information from the docker-compose I was able to successfully run the clustered cockroach sample, but did see the following intermittently
```
roach1_1  | cockroach.cluster(O): W190312 20:31:42.914114 29 server/node.go:886  [n1,summaries] health alerts detected: {Alerts:[{StoreID:1 Category:METRICS Description:queue.replicate.process.failure Value:309}]}
``` 

-----
Below is the diff of the changes I made to the plan.sh and hooks in order to move this to 2.1.6
```
diff --git a/cockroach/hooks/init b/cockroach/hooks/init
index 48aeeb07..9d0730a8 100644
--- a/cockroach/hooks/init
+++ b/cockroach/hooks/init
@@ -11,7 +11,7 @@ done
 cp "{{pkg.svc_config_path}}/ca.key" "{{pkg.svc_var_path}}/certs/private/ca.key"
 chmod 0600 "{{pkg.svc_var_path}}/certs/private/ca.key"

-cockroach cert create-node \
+cockroachoss cert create-node \
   "{{sys.ip}}" \
   "{{sys.hostname}}" \
   "localhost" \
diff --git a/cockroach/hooks/run b/cockroach/hooks/run
index e28fe357..37cf1fb6 100644
--- a/cockroach/hooks/run
+++ b/cockroach/hooks/run
@@ -1,7 +1,7 @@
 #!/bin/sh

 exec 2>&1
-exec cockroach start \
+exec cockroachoss start \
   {{#if cfg.insecure ~}}
   --insecure \
   {{else ~}}
diff --git a/cockroach/plan.sh b/cockroach/plan.sh
index 48081193..cad594d6 100644
--- a/cockroach/plan.sh
+++ b/cockroach/plan.sh
@@ -1,12 +1,12 @@
 pkg_name=cockroach
 pkg_origin=core
-pkg_version=2.0.6
+pkg_version=2.1.6
 pkg_description="CockroachDB is a cloud-native SQL database for building global, scalable cloud services that survive disasters."
 pkg_upstream_url=https://github.com/cockroachdb/cockroach
 pkg_license=('Apache-2.0')
 pkg_maintainer="The Habitat Maintainers <humans@habitat.sh>"
 pkg_source=https://binaries.cockroachdb.com/cockroach-v${pkg_version}.src.tgz
-pkg_shasum=fd1d966542565cbeca6c390db6b7afd98bc239e6e2a4602b3edff973be472097
+pkg_shasum=04a399a619fc898fbd457c60784416e17af22b05f02be101b2fe64c536163eeb
 pkg_dirname="${pkg_name}-v${pkg_version}"
 pkg_deps=(
   core/gcc-libs
@@ -22,6 +22,9 @@ pkg_build_deps=(
   core/gcc
   core/cmake
   core/autoconf
+  core/procps-ng
+  core/git
+  core/diffutils
 )
 pkg_bin_dirs=(bin)
 pkg_exports=(
@@ -48,7 +51,7 @@ do_build() {

 do_install() {
   pushd "./src/github.com/cockroachdb/cockroach" > /dev/null
-  cp -v cockroach bin/* "${pkg_prefix}/bin/"
+  cp -v cockroachoss bin/* "${pkg_prefix}/bin/"
   popd > /dev/null
 }
```

irvingpop commented:
thanks for your help @smacfarlane -  I've rebased to master (which included a version bump to 2.1.3 and dealt with the `cockroachoss` changes) plus your suggestions to go to the latest (2.1.6).

After rebuilding, I observed:
- I did see the permissions errors - something has changed in habitat in the last few months that has changed and is a separate issue.  I removed the docker volume references from the compose example in the README
- I did also see the `queue.replicate.process.failure` message but only once and it was during startup.  I'm not seeing it regularly.

irvingpop commented:
I've pushed `irvingpop/cockroach:2.1.6-20190313052348` to dockerhub, built from this branch, for others to test.

smacfarlane commented:
@irvingpop Thanks for those changes.  This looks good to merge, but I need to do some checking on the license and the naming of the executable produced. 

It appears that this is [dual licensed](https://github.com/cockroachdb/cockroach/blob/master/LICENSE) under Apache2.0 and Cockroachdb Community License, but it's not clear at a quick glance if the latter applies to the opensource builds.

I also need to verify that we can call the executable `cockroach`, there may be a requirement that it be named `cockroachoss`.

smacfarlane commented:
I think we're ok on the license as long as we continue to use `make buildoss`: https://github.com/cockroachdb/cockroach/blob/master/Makefile#L837

Still researching the binary name.

irvingpop commented:
Thanks @smacfarlane I appreciate you digging into that.     At this moment (in master) the binary is already being renamed from `cockroachoss` -> `cockroach` in [this commit](https://github.com/habitat-sh/core-plans/commit/15b0067b52fa6fb5030d5e49be21b646b88b6de0)

If we need to, I can update this PR to undo that change otherwise I'm simply embracing the current state of things

smacfarlane commented:
Looking through the history, the filename was changed as prep work for the base-plans refresh in January but I botched the merge, which is why the filename changed in `run` but not in plan.sh

I haven't been able to find any hard evidence that we need to keep the name as `cockroachoss`, but in the spirit of honoring the upstream default I do think it's important to keep the names of binaries produced intact when possible.  

@irvingpop would you mind updating the plan.sh and hooks to use `cockroachoss`?  Then I think this is good to merge.

irvingpop commented:
Updated @smacfarlane
